blockchain/src/codec.rs

```diff
@@ -0,0 +1,376 @@
+use crate::shortid::ShortIDVec;
+use crate::utreexo::Proof;
+use crate::{
+    Block, BlockHeader, BlockID, BlockTx, GetBlock, GetInventory, GetMempoolTxs, Inventory,
+    MempoolTxs, Message,
+};
+use curve25519_dalek::ristretto::CompressedRistretto;
+use curve25519_dalek::scalar::Scalar;
+use readerwriter::{Decodable, Encodable, ReadError, Reader, WriteError, Writer};
+use std::convert::TryFrom;
+use zkvm::{merkle, Hash, Signature, Tx};
+
+#[repr(u8)]
+enum MessageType {
+    Block = 0,
+    GetBlock = 1,
+    Inventory = 2,
+    GetInventory = 3,
+    MempoolTxs = 4,
+    GetMempoolTxs = 5,
+}
+
+impl TryFrom<u8> for MessageType {
+    type Error = ReadError;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(MessageType::Block),
+            1 => Ok(MessageType::GetBlock),
+            2 => Ok(MessageType::Inventory),
+            3 => Ok(MessageType::GetInventory),
+            4 => Ok(MessageType::MempoolTxs),
+            5 => Ok(MessageType::GetMempoolTxs),
+            _ => Err(ReadError::Custom(
+                format!("unknown message type: {}", value).into(),
+            )),
+        }
+    }
+}
+
+fn read<T>(data: Result<T, ReadError>, label: &'static str) -> Result<T, ReadError> {
+    data.map_err(|err| {
+        ReadError::Custom(
+            match err {
+                ReadError::InvalidFormat => format!("invalid format of {}", label),
+                r => r.to_string(),
+            }
+            .into(),
+        )
+    })
+}
+
+trait ReaderExt: Reader + Sized {
+    fn read_u8_vec(&mut self) -> Result<Vec<u8>, ReadError> {
+        let len = self.read_u32()? as usize;
+        self.read_vec(len)
+    }
+
+    fn read_signature(&mut self) -> Result<Signature, ReadError> {
```

Comment:
little convenience helpers like `read_signature` for external types, not tied to the blockchain encoding business are totally fine to have in the ReaderExt (like ZkVM uses `read_point`, `read_scalar`). But big fat methods for the types we have, like Inventory, or Block - can totally be defined as `impl Decodable` in appropriate places.

Reply:
Removed big fat methods. Added impl Decodable/Encodable